### PR TITLE
Introduce "ondemand" linker sections

### PR DESCRIPTION
**Pull request by npitre**
This is needed when the firmware is significantly bigger than the available
RAM.

This can be seen in action with the ARM64 demand paging support in PR #74129.

### CODE REVIEW

File: Kconfig.zephyr [hunk 1]
```diff
@@ -260,6 +260,20 @@ config LINKER_USE_PINNED_SECTION
 	  Requires that pinned sections exist in the architecture, SoC,
 	  board or custom linker script.
 
+config LINKER_USE_ONDEMAND_SECTION
+	bool "Use Evictable Linker Section"
+	depends on DEMAND_MAPPING
+	depends on !LINKER_USE_PINNED_SECTION
```
Comment (dcpleung) [inline Kconfig.zephyr:266]: Pinned sections can still be used, right?

**Comment by npitre** [comment]
> Pinned sections can still be used, right?

Well, you have to choose. Either you have explicit pinned sections and 
everything else is implicitly evictable, or you have explicit on-demand 
(aka evictable) sections and everything else is implicitly pinned. Both 
tags may coexist of course.

**Comment by dcpleung** [comment]
> Pinned sections can still be used, right?
> Well, you have to choose. Either you have explicit pinned sections and everything else is implicitly evictable, or you have explicit on-demand (aka evictable) sections and everything else is implicitly pinned. Both tags may coexist of course.

Wonder if they should be make mutually exclusive by adding `!LINKER_USE_ONDEMAND_SECTION` to pinned section kconfig.

**Review by dcpleung — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Review by andyross — COMMENTED** [review]
FWIW I do kind of worry that we've now implemented two converse solutions to the same problem...  Clearly the API can be different with how the defaults work, but surely the backend should be the same for "pinned" and "on demand", no?

Not a -1, but sort of a smell complaint.

**Comment by npitre** [comment]
Nothing prevents both strategies from using the same backend. But their 
initial state must differs.

In the explicit pinned model, everything is initially in memory and 
stuff gets evicted when memory is tight. The backing store is used as a 
swap space and is initially unpopu;ated. This assumes the firmware may 
all fit in memory at boot time.

In the explicit on-demand model, most everything is initially in the 
evicted state. The backing store already contains the evicted stuff, 
most likely in read-only mode. This is especially useful when the 
firmware is bigger than available RAM and no XIP is unavailable.

Where both strategies differ the most is in their initial MMU mappings.

**Comment by npitre** [comment]
@andyross @peter-mitsis ping

**Review by andyross — APPROVED** [review]
Oh, never actually +1'd.  Still kind of worry about the bifurcation, but this is clean enough.

**Pull request merged**